### PR TITLE
feat(toast): add toast component

### PR DESCRIPTION
**Pull request by andoulla**
https://recordit.co/c0Oq1NMxwE

### CODE REVIEW

File: sandbox/index.tsx [hunk 1]
```diff
@@ -4,11 +4,17 @@ import * as React from 'react'
 import * as ReactDOM from 'react-dom'
 import { reset } from 'stitches-reset'
 
-import { Box, globalCss } from '../dist'
+import { Box, Button, globalCss, Toast, toast } from '../dist'
```
Comment (andoulla) [inline sandbox/index.tsx:7]: Will be reverting this

File: src/components/toast/Toast.tsx [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+
+const Zoom = cssTransition({
+  enter: 'enter',
+  exit: 'exit',
+  collapseDuration: 200
+})
+
+type ToastProps = React.ComponentPropsWithoutRef<typeof StyledToast>
+
+export const Toast: React.FC<ToastProps> = (props) => {
```
Comment (thomasdigby) [inline src/components/toast/Toast.tsx:126]: Could we update this export to `ToastProvider` as it fits more closely with its purpose? (even though it's not technically a context provider)

File: jest.config.js [hunk 3]
```diff
@@ -6,7 +6,8 @@ module.exports = {
   setupFilesAfterEnv: ['./jest-setup.ts'],
   moduleNameMapper: {
     '~/(.*)': '<rootDir>/src/$1', // matches the path alias in tsconfig.json
-    '^react$': '<rootDir>/node_modules/react'
+    '^react$': '<rootDir>/node_modules/react',
+    '\\.(css|less|sass|scss)$': '<rootDir>/__mocks__/styleMock.js'
```
Comment (thomasdigby) [inline jest.config.js:10]: Can we update this key to `'\\.css$'` just to be super clear what we actually support within the component library, if anyone sneakily adds a scss file I'd be keen for this to fail!

File: src/components/toast/Toast.test.tsx [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+import { Toast, toast } from '.'
+
+describe('Toast component', () => {
+  const message = 'Message'
+
+  it('renders', async () => {
+    render(<Toast />)
+
+    expect(await screen.queryByText(message)).toEqual(null)
```
Comment (hollg) [inline src/components/toast/Toast.test.tsx:13]: Is there a difference between this and `.not.toBeInTheDocument()`?

File: src/components/toast/Toast.tsx [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+    cursor: 'pointer',
+    display: 'flex',
+    fontFamily: '$sans',
+    fontWeight: 400,
+    justifyContent: 'space-between',
+    mb: '$2',
+    minHeight: 60,
+    px: '$2',
+    py: '$3',
+    position: 'relative',
```
Comment (hollg) [inline src/components/toast/Toast.tsx:55]: These seem to all be alphabetical apart from `position` and `overflow`. Could they be made alphabetical too?

File: src/components/toast/Toast.test.tsx (hunk 4 above)
Comment (andoulla) [reply src/components/toast/Toast.test.tsx:13]: queryByText returns a null if it doesn't find the text, but `toBeInTheDocument` expects an HTMLElement so I couldn't use it. I had assumed that the `.not` could cope with the null value but it didn't

File: src/components/toast/Toast.tsx (hunk 2 above)
Comment (andoulla) [reply src/components/toast/Toast.tsx:126]: I could call it ToastContainer to have it match with the terminology of react-toastify.

File: src/components/toast/Toast.test.tsx (hunk 4 above)
Comment (hollg) [reply src/components/toast/Toast.test.tsx:13]: We have lots of tests with assertions like this: 
```
// AlertDialog.test.tsx
expect(await screen.queryByText('CONTENT')).not.toBeInTheDocument()

// Button.test.tsx
expect(await screen.queryByText('BUTTON')).not.toBeVisible()
```

What's the difference that stops it from working in this case? I think it's worth using if possible as it's more descriptive about what's going on (and doesn't require knowledge about the circumstances in which `testing-library` functions return `null`).

Comment (andoulla) [reply src/components/toast/Toast.test.tsx:13]: I'm not really sure  why it doesn't work. Actually none of these work:
```
// expect(await screen.queryByText(message)).not.toBeVisible()
    // expect(await screen.queryByText(message)).not.toBeInDocument()
    // expect(await screen.queryByText(message)).not.toContainElement()
```
But from various different blogs I know I can switch this to 
`expect(await screen.queryByText(message)).toBeFalsy()`
which might be a bit more readable

File: src/components/toast/Toast.tsx [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+  '.Toastify__toast--success': {
+    bg: '$success'
+  },
+  '.Toastify__toast--warning': {
+    bg: '$warning'
+  },
+  '.Toastify__toast--error': {
+    bg: '$danger'
+  },
+  '.Toastify__progress-bar--animated': {
```
Comment (hollg) [inline src/components/toast/Toast.tsx:82]: Do we need this if we're hiding the progress bar?

File: yarn.lock [hunk 7]
```diff
@@ -9631,9 +9650,9 @@ saxes@^5.0.0:
     xmlchars "^2.2.0"
 
 scheduler@^0.20.1:
-  version "0.20.1"
-  resolved "https://registry.yarnpkg.com/scheduler/-/scheduler-0.20.1.tgz#da0b907e24026b01181ecbc75efdc7f27b5a000c"
-  integrity sha512-LKTe+2xNJBNxu/QhHvDR14wUXHRQbVY5ZOYpOGWRzhydZUqrLb2JBvLPY7cAqFmqrWuDED0Mjk7013SZiOz6Bw==
+  version "0.20.2"
+  resolved "https://registry.npmjs.org/scheduler/-/scheduler-0.20.2.tgz#4baee39436e34aa93b4874bddcbf0fe8b8b50e91"
```
Comment (hollg) [inline yarn.lock:9654]: Why have these registries changed? Is it an issue? 🤷🏻‍♂️

Comment (andoulla) [reply yarn.lock:9654]: I have no idea why they have changed. I will revert them. Good catch

File: src/components/toast/Toast.tsx (hunk 6 above)
Comment (andoulla) [reply src/components/toast/Toast.tsx:82]: no we don't, I have removed everything else missed this one. Thanks

**Review by hollg — APPROVED** [review]

**Comment by thomasdigby** [comment]
@andoulla Can you share some images of the toast? The video you recorded looks different to the toasts we have in the public app and it doesn't look like the styles in your CSS have applied

File: sandbox/index.tsx [hunk 8]
```diff
@@ -5,6 +5,7 @@ import * as ReactDOM from 'react-dom'
 import { reset } from 'stitches-reset'
 
 import { Box, globalCss } from '../dist'
+
```
Comment (thomasdigby) [inline sandbox/index.tsx:8]: 🙏 😆

File: src/components/toast/Toast.mdx [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+This is the preferred user feedback way as it is not intrusive and it simply provides some state information to the user without waiting for feedback, ie: the information was saved. When feedback is required consider using the `AlertDialog` component.
+
+As this component uses `react-toastify` please read more information about the underlying behaviour at [react-toastify](https://fkhadra.github.io/react-toastify/introduction/)
+
+```jsx preview
+import { ToastProvider, toast } from '@atom-learning/components'
+
+const handleSuccess = () => toast.success('message')
+
+<ToastProvider />
```
Comment (thomasdigby) [inline src/components/toast/Toast.mdx:19]: Can we update this example to include a little more boilerplate like `AlertDialog` just to clarify the structure and components required to make it work. e.g.

```jsx
import { ToastProvider, Button, toast } from '@atom-learning/components'

const Component = () => (
  <Button onClick={() => toast.success('message')}>
    Delete school
  </Button>
)

const App = () => (
  <ToastProvider>
    <Component />
  </ToastProvider>
)
```

File: src/components/toast/Toast.tsx [hunk 10]
```diff
@@ -0,0 +1,124 @@
...
+  },
+  '.Toastify__toast--warning': {
+    bg: '$warning'
+  },
+  '.Toastify__toast--error': {
+    bg: '$danger'
+  },
+  '.Toastify__close-button': {
+    background: 'unset',
+    border: 'unset',
```
Comment (thomasdigby) [inline src/components/toast/Toast.tsx:79]: I've started using `all: unset` more for these use-cases where we need to remove all styling from a button.

File: src/components/toast/Toast.tsx [hunk 11]
```diff
@@ -0,0 +1,124 @@
+import 'react-toastify/dist/ReactToastify.css'
```
Comment (thomasdigby) [inline src/components/toast/Toast.tsx:1]: Does this include all of the `react-toastify` default styles? I think you may need to import `'react-toastify/dist/ReactToastify.minimal.css'` instead as we're providing the toast styles

File: sandbox/index.tsx (hunk 8 above)
Comment (andoulla) [reply sandbox/index.tsx:8]: I'm not sure what I need to change?

File: src/components/toast/Toast.tsx [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+import * as React from 'react'
+import { cssTransition, ToastContainer } from 'react-toastify'
+
+import { keyframes, styled } from '~/stitches'
+
+const toastSize = '44px'
+const toastOffset = '32px'
```
Comment (thomasdigby) [inline src/components/toast/Toast.tsx:9]: These values could use the sizing scale e.g.

```js
const toastSize = '$sizes$4'
const toastOffset = '$sizes$5'
```

File: src/components/toast/Toast.tsx [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+    borderRadius: '$0',
+    boxShadow: '$1',
+    color: 'white',
+    cursor: 'pointer',
+    display: 'flex',
+    fontFamily: '$sans',
+    fontWeight: 400,
+    justifyContent: 'space-between',
+    mb: '$2',
+    minHeight: 60,
```
Comment (thomasdigby) [inline src/components/toast/Toast.tsx:47]: This could be `$5` to get `64px` from our sizing scale

File: sandbox/index.tsx (hunk 8 above)
Comment (thomasdigby) [reply sandbox/index.tsx:8]: Oh nothing, I was just thanking you for restoring the space, I'm glad someone wants to ensure we have a space following the imports as much as I do 😅

File: src/components/toast/Toast.tsx (hunk 12 above)
Comment (andoulla) [reply src/components/toast/Toast.tsx:9]: these are outside the styled function, can I use them like this?

Comment (thomasdigby) [reply src/components/toast/Toast.tsx:9]: They're only used within stitches `styled` function so they should be parsed correctly at runtime... I'm not sure when stitches access its token values but it _should_ parse the CSS value as a string and the tokens should be replaced inline. e.g. when stitches sees `translate3d(0, -calc(${toastSize + toastOffset}), 0)` it should evaluate that statement and see `translate3d(0, -calc($sizes$4 + $sizes$5), 0)`

**Review by andoulla — COMMENTED** [review]
@thomasdigby these are the latest ones, including the sizing changes you recommended
image

**Comment by andoulla** [comment]
> @thomasdigby these are the latest ones, including the sizing changes you recommended
> image

They could be a bit slimmer if we would like

File: src/components/toast/Toast.mdx (hunk 9 above)
Comment (andoulla) [reply src/components/toast/Toast.mdx]: I just stole your suggestion

File: src/components/toast/Toast.mdx [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+
+This is the preferred user feedback way as it is not intrusive and it simply provides some state information to the user without waiting for feedback, ie: the information was saved. When feedback is required consider using the `AlertDialog` component.
+
+This component is a wrapper for `react-toastify`
+
+```jsx preview
+import { Button, ToastProvider, toast } from '@atom-learning/components'
+
+const Component = () => (
+  <Button onClick={() => toast.success('message')}>Delete school</Button>
```
Comment (thomasdigby) [inline src/components/toast/Toast.mdx:18]:
```suggestion
  <Button onClick={() => toast.success('message')}>Click for toast</Button>
```

**Comment by andoulla** [comment]
> > @thomasdigby these are the latest ones, including the sizing changes you recommended
> > image
> 
> They could be a bit slimmer if we would like

Updated versions (reduced margins and height, used a lighter box-shadow)
image


image

**Review by thomasdigby — APPROVED** [review]
🍞

**Pull request merged**